### PR TITLE
Prototype MVC changes for next auth iteration

### PR DESCRIPTION
**Pull request by HaoK**
- Prototype switch to using common AuthorizeAttribute, still need to rationalize the AuthorizationFilterAttribute, this change focused on removing the MVC AuthorizeAttribute and preserving the exact same functionality (tests all pass)

Should what I call AuthorizationFilter, really be called AuthorizeFilter since it basically is a filter that was created from an Authorize attribute?

cc @yishaigalatzer @harshgMSFT @rynowak @blowdart @lodejard

### CODE REVIEW

**Comment by blowdart** [comment]
Naming is hard. But yea, AuthorizeFilter seems better.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs [hunk 1]
```diff
@@ -1,46 +1,25 @@
...
 using System.Threading.Tasks;
-using Microsoft.AspNet.Mvc.Core;
 using Microsoft.AspNet.Security;
 using Microsoft.Framework.DependencyInjection;
 
 namespace Microsoft.AspNet.Mvc
 {
-    public class AuthorizeAttribute : AuthorizationFilterAttribute
+    // REVIEW: figure out how to propery rationalize this with AuthorizationFilterAttribute
+    public class AuthorizationFilter : IAsyncAuthorizationFilter, IAuthorizationFilter, IOrderedFilter
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs]: you don't need to implement both AsyncAuthorizationFilter and `IAuthorizationFilter` it was there for overriding the property but probably no longer necessary?

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs [hunk 2]
```diff
@@ -53,11 +32,11 @@ public string Roles
             var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
 
             // Build a policy for the requested roles if specified
-            if (_rolesSplit != null)
+            if (Roles != null && Roles.Any())
             {
                 var rolesPolicy = new AuthorizationPolicyBuilder();
-                rolesPolicy.RequiresRole(_rolesSplit);
-                if (!await authService.AuthorizeAsync(rolesPolicy.Build(), httpContext, context))
+                rolesPolicy.RequiresRole(Roles.ToArray());
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs]: Lets not `ToArray` on every call

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs (hunk 1 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs]: Removed sync version, only implements AsyncAuthorizationFilter now

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs (hunk 2 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizationFilter.cs]: Fair enough, I'll add an overload to Policy builder which also allows `IEnumerable<string>`

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+namespace Microsoft.AspNet.Mvc
+{
+    // REVIEW: figure out how to propery rationalize this with AuthorizationFilterAttribute
+    public class AuthorizeFilter : IAsyncAuthorizationFilter, IOrderedFilter
+    {
+        public int Order { get; set; }
+
+        public string Policy { get; set; }
+
+        public IEnumerable<string> Roles { get; set; }
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Make this an `IList<string>` so no ToArray is required later.

Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: The issue is that the AuthZ apis only took params string[], so IList doesn't help, I added the overloads that accept IEnumerable

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs [hunk 4]
```diff
@@ -256,6 +257,15 @@ private bool IsIDisposableMethod(MethodInfo methodInfo, TypeInfo typeInfo)
...
             AddRange(actionModel.Filters, attributes.OfType<IFilter>());
 
+            foreach (var authorizeAttribute in attributes.OfType<AuthorizeAttribute>())
+            {
+                actionModel.Filters.Add(new AuthorizeFilter
+                {
+                    Policy = authorizeAttribute.Policy,
+                    Roles = authorizeAttribute.Roles?.Split(','),
+                });
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: We need to see the same in `ControllerModelBuilder`. This code only looks at attributes on the method.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+#pragma warning disable 1998
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: why? looks like this method does have async code.

File: test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs [hunk 6]
```diff
@@ -1835,15 +1835,15 @@ private struct SampleStruct
...
+            var AuthorizeFilter = new Mock<IAuthorizationFilter>(MockBehavior.Strict);
+            AuthorizeFilter
                 .Setup(f => f.OnAuthorization(It.IsAny<AuthorizationContext>()))
                 .Callback<AuthorizationContext>((c) =>
                 {
                     c.Result = _result;
                 });
 
-            var invoker = CreateInvoker(new IFilter[] { authorizationFilter.Object, resourceFilter.Object, });
+            var invoker = CreateInvoker(new IFilter[] { AuthorizeFilter.Object, resourceFilter.Object, });
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ControllerActionInvokerTest.cs]: revert these - find and replace fail

File: test/Microsoft.AspNet.Mvc.Core.Test/DefaultActionDescriptorCollectionProviderLoggingTest.cs [hunk 7]
```diff
@@ -72,7 +73,7 @@ public void ControllerDiscovery()
             Assert.Empty(controllerModelValues.AttributeRoutes);
             Assert.Empty(controllerModelValues.RouteConstraints);
             Assert.NotEmpty(controllerModelValues.Attributes);
-            Assert.Single(controllerModelValues.Filters);
+            Assert.Empty(controllerModelValues.Filters);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/DefaultActionDescriptorCollectionProviderLoggingTest.cs]: This shouldn't have changed - fix the `ControllerModelBuilder` and this will be the same as before

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeAttributeTests.cs [hunk 8]
```diff
@@ -15,7 +15,7 @@ public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
         public async Task Invoke_ValidClaimShouldNotFail()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeAttributeTests.cs]: rename file

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Security;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    // REVIEW: figure out how to propery rationalize this with AuthorizationFilterAttribute
+    public class AuthorizeFilter : IAsyncAuthorizationFilter, IOrderedFilter
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: needs docs

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: Need to see a unit test for this and also for controllermodelbuilder when that's fixed too.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 5 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: This was the old AuthorizeAttribute code, I just copied it over, I can remove it

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeAttributeTests.cs (hunk 8 above)
Comment (HaoK) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeAttributeTests.cs]: Renamed

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 4 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: Added unit tests

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs [hunk 10]
```diff
@@ -321,6 +322,27 @@ public void GetActions_ConventionallyRoutedAction_WithHttpConstraints()
         }
 
         [Fact]
+        public void GetActions_AuthorizeFiltersAreCombinedWhenOverriden()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: This isn't really accurate is it? 2 filters are created and both with run so how are they combined?

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultControllerModelBuilderTest.cs [hunk 11]
```diff
@@ -7,6 +7,8 @@
 using Xunit;
 using System;
 using System.Threading.Tasks;
+using Microsoft.AspNet.Security;
+using System.Linq;
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultControllerModelBuilderTest.cs]: sort

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs [hunk 12]
```diff
@@ -9,13 +9,13 @@
 
 namespace Microsoft.AspNet.Mvc.Core.Test
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs]: namespace should match product code

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTestsBase.cs [hunk 13]
```diff
@@ -12,7 +12,7 @@
 
 namespace Microsoft.AspNet.Mvc.Core.Test
 {
-    public class AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTestsBase
     {
         protected AuthorizationContext GetAuthorizationContext(Action<ServiceCollection> registerServices, bool anonymous = false)
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTestsBase.cs]: This is wierd and only used in one place. I know it's not a mess that you made but you can just move this method to the other class and delete this?

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNet.Security;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// An implementation of <see cref="IAsyncAuthorizationFilter"/> which corresponds to [Authorize]
+    /// </summary>
+    public class AuthorizeFilter : IAsyncAuthorizationFilter, IOrderedFilter
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Sorry, wasn't clear, need docs on all of the public/protected stuff. If it's an interface method you can just use `/// <inheritdoc />`

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: I don't really grok this comment or the response. I don't see a `ToArray()` so I'm assuming there's no issue here?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: :+1:

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs [hunk 15]
```diff
@@ -9,13 +9,13 @@
...
 {
-    public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTests : AuthorizeFilterTestsBase
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTests.cs]: `AuthorizeFilterTest` (yeah I know it was wrong before too)

**Comment by rynowak** [comment]
:watch: looking pretty close to done

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 3 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Yeah I fixed it in security itself by adding overloads that take IEnumerable directly, so there no longer is an issue with ToArray

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs (hunk 10 above)
Comment (HaoK) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: Poor test name, the intent was that both base and derived authorize filters show up, as opposed to getting only one.  I'll rename it to BaseAuthorizeFiltersAreStillValidWhenOverriden

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        /// <inheritdoc />
+        public int Order { get; set; }
+
+        /// <summary>
+        /// Policy name used for authorization
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: nit: all summary should end with a .

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+                ? await authService.AuthorizeAsync(httpContext.User, context, BuildAnyAuthorizedUserPolicy())
+                : await authService.AuthorizeAsync(httpContext.User, context, Policy);
+            if (!authorized)
+            {
+                Fail(context);
+            }
+        }
+
+        /// <summary>
+        /// Returns true if there is an IAllowAnonymous filter
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: <see cref="IAllowAnonymous" />

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        /// <returns></returns>
+        protected virtual bool HasAllowAnonymous([NotNull] AuthorizationContext context)
+        {
+            return context.Filters.Any(item => item is IAllowAnonymous);
+        }
+
+        /// <inheritdoc />
+        protected virtual void Fail([NotNull] AuthorizationContext context)
+        {
+            context.Result = new HttpStatusCodeResult(401);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Is this the recommended way? should there be a Fail and Success on the service/context itself? and the authorization gets to 401 or 403 depending on what failed? I know this is doing exactly what the code before does, but wondering if the pattern is going to be this or going to change.

File: test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTest.cs [hunk 19]
```diff
@@ -1,21 +1,75 @@
...
 using Moq;
 using Xunit;
 
-namespace Microsoft.AspNet.Mvc.Core.Test
+namespace Microsoft.AspNet.Mvc.Test
 {
-    public class AuthorizeAttributeTests : AuthorizeAttributeTestsBase
+    public class AuthorizeFilterTest
     {
+        private AuthorizationContext GetAuthorizationContext(Action<ServiceCollection> registerServices, bool anonymous = false)
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/Filters/AuthorizeFilterTest.cs]: nit: can we move this to end of the file? that is what we have been following.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 18 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: also use ```Microsoft.AspNet.WebUtilities.StatusCodes.Status401Unauthorized ```  instead of 401

**Comment by harshgMSFT** [comment]
:watch:

**Comment by HaoK** [comment]
Updated with latest iteration changes, and PR feedback:
- AuthorizeFilter expects to get a combined policy (union of all the authorize attributes)
- Returns 401 if anonymous, 403 if authorization service fails
- Updated functional to be less verby

**Comment by borgdylan** [comment]
Shoudldn't the AuthorizeFilter support a policy less constructor? That would make basic usage easier rather than having to define policies in DI for a simple case.

**Comment by HaoK** [comment]
@borgdylan The new authorize filter expects the combined policies, if there is not supposed to be any authorization done at all (which is the empty policy less constructor case) there shouldn't be an authorize filter at all.  This is also a side effect of the decision to make an empty policy with no requirements always fail auth.

**Comment by borgdylan** [comment]
But will there be a way to make a policy that allows anyone that presents valid credentials to pass successfully?

**Comment by HaoK** [comment]
Sure, policy.RequireAnyAuthenticatedUser()

**Comment by HaoK** [comment]
@harshgMSFT @yishaigalatzer @rynowak Is this PR ready for checkin?  There's another big wave of authN changes coming, so I want to get this iteration of AuthZ in before tackling that one.

**Comment by lodejard** [comment]
:shipit: ..?

File: samples/MvcSample.Web/Startup.cs [hunk 20]
```diff
@@ -107,6 +107,19 @@ public void Configure(IApplicationBuilder app)
                         options.Filters.Add(typeof(PassThroughAttribute), order: 17);
                         options.AddXmlDataContractSerializerFormatter();
                     });
+                    services.ConfigureAuthorization(auth =>
+                    {
+                        auth.AddPolicy("CanViewPage",
```
Comment (harshgMSFT) [inline samples/MvcSample.Web/Startup.cs]: nit: formatting : all parameters should be on top of one another. 
something like 
```c#
    auth.AddPolicy(
        "CanViewAnything",
        new AuthorizationPolicyBuilder().RequiresClaim("Permission", "CanViewAnything")
            .Build());
```

File: samples/MvcSample.Web/Startup.cs [hunk 21]
```diff
@@ -107,6 +107,19 @@ public void Configure(IApplicationBuilder app)
...
+                            new AuthorizationPolicyBuilder()
+                                .RequiresClaim("Permission", "CanViewPage", "CanViewAnything").Build());
+                        auth.AddPolicy("CanViewAnything",
+                            new AuthorizationPolicyBuilder()
+                                .RequiresClaim("Permission", "CanViewAnything").Build());
+                        // This policy basically requires that the auth type is present
```
Comment (harshgMSFT) [inline samples/MvcSample.Web/Startup.cs]: nit: blank line above this

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs [hunk 22]
```diff
@@ -16,6 +18,13 @@ namespace Microsoft.AspNet.Mvc.ApplicationModels
     /// </summary>
     public class DefaultActionModelBuilder : IActionModelBuilder
     {
+        private readonly AuthorizationOptions _authorizationOptions;
+
+        public DefaultActionModelBuilder(IOptions<AuthorizationOptions> options = null)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: docs. And we do not use default values. Why is this needed?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 23]
```diff
@@ -20,15 +22,17 @@ public class DefaultControllerModelBuilder : IControllerModelBuilder
...
         private readonly IActionModelBuilder _actionModelBuilder;
         private readonly ILogger _logger;
+        private readonly AuthorizationOptions _authorizationOptions;
 
         /// <summary>
         /// Creates a new <see cref="DefaultControllerModelBuilder"/>.
         /// </summary>
         /// <param name="actionModelBuilder">The <see cref="IActionModelBuilder"/> used to create actions.</param>
-        public DefaultControllerModelBuilder(IActionModelBuilder actionModelBuilder, ILoggerFactory loggerFactory)
+        public DefaultControllerModelBuilder(IActionModelBuilder actionModelBuilder, ILoggerFactory loggerFactory, IOptions<AuthorizationOptions> options = null)
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: break the line so that length is <120

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: as above, do not use default parameters.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// An implementation of <see cref="IAsyncAuthorizationFilter"/>
+    /// </summary>
+    public class AuthorizeFilter : IAsyncAuthorizationFilter
+    {
+        /// <summary>
+        /// Authorize filter for a specific policy
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: All the docs should finish with a full stop ```.```

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        /// <param name="policy"></param>
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: missing docs.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        public AuthorizationPolicy Policy { get; private set; }
+
+        /// <inheritdoc />
+        public virtual async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // Build a ClaimsPrincipal with the Policy's required authentication types
+            if (Policy.ActiveAuthenticationTypes != null && Policy.ActiveAuthenticationTypes.Any())
+            {
+                var results = await context.HttpContext.AuthenticateAsync(Policy.ActiveAuthenticationTypes);
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: can the results be null?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs (hunk 22 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultActionModelBuilder.cs]: Made it so I didn't have to update all the unit tests, but I can remove it

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+
+            // Allow Anonymous skips all authorization
+            if (context.Filters.Any(item => item is IAllowAnonymous))
+            {
+                return;
+            }
+
+            var httpContext = context.HttpContext;
+            var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
+            // Note: Default Anonymous User is new ClaimsPrincipal(new ClaimsIdentity())
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: nit: a newline above this.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+            }
+
+            // Allow Anonymous skips all authorization
+            if (context.Filters.Any(item => item is IAllowAnonymous))
+            {
+                return;
+            }
+
+            var httpContext = context.HttpContext;
+            var authService = httpContext.RequestServices.GetRequiredService<IAuthorizationService>();
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Instead of calling getRequired service here(i.e within the filter), the pattern that we follow is to have an implementation of ```IFilterFactory``` and create the instance of the filter there. So in this case we would have something like 
```c# 
public class AuthorizeFilterFactory : IFilterFactory

```
and then create the filter in its create method ( the advantage here is that create method takes in an IServiceProvider which can be used to resolve the authorization service. ).

Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Can you point me to an example in the current sources?

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs [hunk 29]
```diff
@@ -126,6 +130,12 @@ protected virtual ControllerModel CreateControllerModel([NotNull] TypeInfo typeI
             AddRange(controllerModel.Filters, attributes.OfType<IFilter>());
             AddRange(controllerModel.RouteConstraints, attributes.OfType<IRouteConstraintProvider>());
 
+            var policy = AuthorizationPolicy.Combine(_authorizationOptions, attributes.OfType<AuthorizeAttribute>());
+            if (policy != null)
+            {
+                controllerModel.Filters.Add(new AuthorizeFilter(policy));
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: Instead of creating the AuthorizeFilter here, probably we should create a filter factory. That would automatically create the filter at a later point.  Also creating the factory here will enable you to not take here in ```DefaultControllerModelBuilder``` and ```DefaultActionModelBuilder```. Please see the comment on AuthorizeFilter as well.

File: test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs [hunk 30]
```diff
@@ -646,8 +672,24 @@ public void GetActions_MixedHttpVerbsAndRoutes_MultipleEmptyAndNonEmptyVerbs()
             Assert.Equal<string>(new string[] { "GET" }, action.HttpMethods);
         }
 
+        private static DefaultActionModelBuilder CreateTestDefaultActionModelBuilder(AuthorizationOptions authOptions = null)
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.Core.Test/ApplicationModel/DefaultActionModelBuilderTest.cs]: line breaks.

File: test/WebSites/XmlFormattersWebSite/XmlFormattersWebSite.xproj [hunk 31]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (harshgMSFT) [inline test/WebSites/XmlFormattersWebSite/XmlFormattersWebSite.xproj]: revert.

File: test/WebSites/FiltersWebSite/Startup.cs [hunk 32]
```diff
@@ -9,6 +10,25 @@
...
+    public static class Operations
+    {
+        public static OperationAuthorizationRequirement Edit = new OperationAuthorizationRequirement { Name = "Edit" };
+        public static OperationAuthorizationRequirement Create = new OperationAuthorizationRequirement { Name = "Create" };
+        public static OperationAuthorizationRequirement Delete = new OperationAuthorizationRequirement { Name = "Delete" };
+    }
+
+    public class ManagerHandler : AuthorizationHandler<OperationAuthorizationRequirement>
+    {
+        public override Task HandleAsync(Microsoft.AspNet.Security.AuthorizationContext context, OperationAuthorizationRequirement requirement)
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/Startup.cs]: line breaks.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 28 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: look at https://github.com/aspnet/Mvc/blob/b98de5c5a305718487f616a4f5f080fb2a59be78/src/Microsoft.AspNet.Mvc.Core/Filters/ResponseCacheAttribute.cs#L88

File: test/WebSites/FiltersWebSite/Startup.cs [hunk 33]
```diff
@@ -9,6 +10,25 @@
...
+        {
+            if (context.User.HasClaim("Manager", "yes"))
+            {
+                context.Succeed(requirement);
+            }
+            return Task.FromResult(0);
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/Startup.cs]: nit: use ```Task.FromResult(true);```

File: test/WebSites/FiltersWebSite/Startup.cs [hunk 34]
```diff
@@ -9,6 +10,25 @@
...
+        public static OperationAuthorizationRequirement Edit = new OperationAuthorizationRequirement { Name = "Edit" };
+        public static OperationAuthorizationRequirement Create = new OperationAuthorizationRequirement { Name = "Create" };
+        public static OperationAuthorizationRequirement Delete = new OperationAuthorizationRequirement { Name = "Delete" };
+    }
+
+    public class ManagerHandler : AuthorizationHandler<OperationAuthorizationRequirement>
+    {
+        public override Task HandleAsync(Microsoft.AspNet.Security.AuthorizationContext context, OperationAuthorizationRequirement requirement)
+        {
+            if (context.User.HasClaim("Manager", "yes"))
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/Startup.cs]: nit: named parameters. Will be easier to review and read.

File: test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs [hunk 35]
```diff
@@ -26,9 +25,12 @@ public class AuthorizeBasicMiddleware : AuthenticationMiddleware<BasicOptions>
         public AuthorizeBasicMiddleware(
             RequestDelegate next, 
             IServiceProvider services, 
-            IOptions<BasicOptions> options) : 
-                base(next, services, options, null)
-        { }
+            IOptions<BasicOptions> options,
+            string authType) : 
+                base(next, services, options,
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs]: if you are splitting the parameters, have all of them on a newline.

File: test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs [hunk 36]
```diff
@@ -15,10 +16,8 @@ public class BasicOptions : AuthenticationOptions
     {
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs]: This file contains 3 classes, which is something we do not follow. (even thought this is not directly related to your change ) could you please split this file into 3?

File: test/WebSites/FiltersWebSite/Startup.cs [hunk 37]
```diff
@@ -9,6 +10,25 @@
 
 namespace FiltersWebSite
 {
+    public static class Operations
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/Startup.cs]: separate files for separate classes.

File: test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs [hunk 38]
```diff
@@ -49,12 +51,13 @@ protected override void ApplyResponseGrant()
         protected override AuthenticationTicket AuthenticateCore()
```
Comment (harshgMSFT) [inline test/WebSites/FiltersWebSite/AuthorizeBasicMiddleware.cs]: should this class be worried about handling a response challenge?

**Comment by harshgMSFT** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs (hunk 29 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: Filed https://github.com/aspnet/Mvc/issues/2022 to track this for later

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: I don't understand the comment here. `IFilterFactory` is only useful when when a filter needs indirection (replaceable) or needs access to request services. 

If you made this `IFilterFactory` you'd still need code here because `AuthorizeAttribute` cannot implement `IFilterFactory` itself.

Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/ApplicationModels/DefaultControllerModelBuilder.cs]: the filter factory is useful since the AuthorizeFilter can directly get the service via a factory. Agreed that this part of the code will still have to do what its doing today.

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 26 above)
Comment (HaoK) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: Added null guard

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs [hunk 39]
```diff
@@ -0,0 +1,62 @@
...
+        /// <inheritdoc />
+        public virtual async Task OnAuthorizationAsync([NotNull] AuthorizationContext context)
+        {
+            // Build a ClaimsPrincipal with the Policy's required authentication types
+            if (Policy.ActiveAuthenticationTypes != null && Policy.ActiveAuthenticationTypes.Any())
+            {
+                var results = await context.HttpContext.AuthenticateAsync(Policy.ActiveAuthenticationTypes);
+                if (results != null)
+                {
+                    context.HttpContext.User = new ClaimsPrincipal(results.Where(r => r.Identity != null).Select(r => r.Identity));
```
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: fix the long line

File: src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs (hunk 28 above)
Comment (yishaigalatzer) [inline src/Microsoft.AspNet.Mvc.Core/Filters/AuthorizeFilter.cs]: This doesn't add much value. I'm suggesting won't fix

**Comment by HaoK** [comment]
Merged, will address any leftover feedback with the next authZ/N iteration update 94ef6cf8fa9e62bca108ef20161f77ce45789a04
